### ISSUE TITLE
Allow unsupported syntax, like record classes

### ISSUE THREAD

**Issue by runfalk** [checkstyle/checkstyle#8023]
In a new project we are using Java 14 Record classes (preview feature). The problem is that checkstyle crashes when parsing record classes with methods like:

```java
public record Person(
        String firstName,
        String lastName,
        int age,
        String address,
        Date birthday
){
    public Person{
      // record ctor
    }

    public String fullName(){
        return firstName + " " + lastName;
    }
}
```

It sort of works with an error when there are no methods.

While I appreciate that you may not want to add support for a feature that is subject to change, it would be very helpful to have some way of skipping checks for new syntax. I'm aware of the possibility to exclude files using regex, but the great thing about record classes is that you can embed them in classes without the number of lines blowing up.

Preferably I'd like some way of ignoring the record classes but still check the rest of the code in the file.

**Comment by rnveach** [comment]
> Preferably I'd like some way of ignoring the record classes but still check the rest of the code in the file.

Most likely this won't be possible without excluding the entire file. To be able to skip a record class, we need to know when it starts and when it ends. It is impossible to make this determination without parsing out every little bit and this will require the full implementation to be supported.

You will have to exclude the files from being processed until we add support for this preview feature, if the issue is approved.

**Comment by runfalk** [comment]
Maybe this could be supported by allowing plugins to preprocess the files before checkstyle look at them?

**Comment by romani** [comment]
Plugins should make decision to run validation or not, non of existing support such approach.
I hope we will be able to provide support  till end of summer.

**Comment by nmancus1** [comment]
@pbludov @romani How would this be?
```
➜  checkstyle git:(testing-for-records) ✗ cat ~/Desktop/testing-for-records/TestClass.java
public record Person(
        String firstName,
        String lastName,
        int age,
        String address,
        Date birthday
){
    public String fullName(){
        return firstName + " " + lastName;
    }
}                                                               
➜  checkstyle git:(testing-for-records) ✗ java -jar target/checkstyle-8.33-SNAPSHOT-all.jar -t ~/Desktop/testing-for-records/TestClass.java 
RECORD_DEF -> RECORD_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_RECORD -> record [1:7]
|--IDENT -> Person [1:14]
|--PARAMETERS -> PARAMETERS [2:8]
|   |--PARAMETER_DEF -> PARAMETER_DEF [2:8]
|   |   |--MODIFIERS -> MODIFIERS [2:8]
|   |   |--TYPE -> TYPE [2:8]
|   |   |   `--IDENT -> String [2:8]
|   |   `--IDENT -> firstName [2:15]
|   |--COMMA -> , [2:24]
|   |--PARAMETER_DEF -> PARAMETER_DEF [3:8]
|   |   |--MODIFIERS -> MODIFIERS [3:8]
|   |   |--TYPE -> TYPE [3:8]
|   |   |   `--IDENT -> String [3:8]
|   |   `--IDENT -> lastName [3:15]
|   |--COMMA -> , [3:23]
|   |--PARAMETER_DEF -> PARAMETER_DEF [4:8]
|   |   |--MODIFIERS -> MODIFIERS [4:8]
|   |   |--TYPE -> TYPE [4:8]
|   |   |   `--LITERAL_INT -> int [4:8]
|   |   `--IDENT -> age [4:12]
|   |--COMMA -> , [4:15]
|   |--PARAMETER_DEF -> PARAMETER_DEF [5:8]
|   |   |--MODIFIERS -> MODIFIERS [5:8]
|   |   |--TYPE -> TYPE [5:8]
|   |   |   `--IDENT -> String [5:8]
|   |   `--IDENT -> address [5:15]
|   |--COMMA -> , [5:22]
|   `--PARAMETER_DEF -> PARAMETER_DEF [6:8]
|       |--MODIFIERS -> MODIFIERS [6:8]
|       |--TYPE -> TYPE [6:8]
|       |   `--IDENT -> Date [6:8]
|       `--IDENT -> birthday [6:13]
`--OBJBLOCK -> OBJBLOCK [7:1]
    |--LCURLY -> { [7:1]
    |--METHOD_DEF -> METHOD_DEF [8:4]
    |   |--MODIFIERS -> MODIFIERS [8:4]
    |   |   `--LITERAL_PUBLIC -> public [8:4]
    |   |--TYPE -> TYPE [8:11]
    |   |   `--IDENT -> String [8:11]
    |   |--IDENT -> fullName [8:18]
    |   |--LPAREN -> ( [8:26]
    |   |--PARAMETERS -> PARAMETERS [8:27]
    |   |--RPAREN -> ) [8:27]
    |   `--SLIST -> { [8:28]
    |       |--LITERAL_RETURN -> return [9:8]
    |       |   |--EXPR -> EXPR [9:31]
    |       |   |   `--PLUS -> + [9:31]
    |       |   |       |--PLUS -> + [9:25]
    |       |   |       |   |--IDENT -> firstName [9:15]
    |       |   |       |   `--STRING_LITERAL -> " " [9:27]
    |       |   |       `--IDENT -> lastName [9:33]
    |       |   `--SEMI -> ; [9:41]
    |       `--RCURLY -> } [10:4]
    `--RCURLY -> } [11:0]


```

**Comment by pbludov** [comment]
I like the idea of using some kind of "record component def" instead of `PARAMETER_DEF`.
These parameters are also fields. It is better to create a new token type here.

As for the record body, we have one "OBJBLOCK" for all types, including enums. And it works well.

So I'd like to see two new token types in the AST: `RECORD_COMPONENT_DEF` and `RECORD_CTOR_DEF` to support parameteless ctors in records.

**Comment by nmancus1** [comment]
@pbludov Ok, sounds great. I am going to start a new issue for the full tree/ full support, referencing your comment above.  To close **this** issue, we would just need a tree like this, right?

```
RECORD_DEF -> RECORD_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_RECORD -> record [1:7]
`--IDENT -> Person [1:14]

```

**Comment by pbludov** [comment]
> @pbludov Ok, sounds great. I am going to start a new issue for the full tree/ full support, referencing your comment above. To close **this** issue, we would just need a tree like this, right?

yes

**Comment by nmancus1** [comment]
@pbludov @esilkensen I have made a preliminary attempt at the record grammar, and ran a check behavior regression report on three projects, that can be found here: https://nmancus1.github.io/diff_5_13_initial-records-support/  

I think the most challenging part of implementing records will be working around variable declarations such as `LogRecord record = log();`  I have tried a few different semantic predicates and actions to try and avoid this, but haven't had much luck yet.  I think I may need to examine the ordering of rules next.

**Comment by nmancus1** [comment]
Also, I wanted to post some links to articles I found about records here, to make it easy for mentors to find any information and examples of records implementation.

1. A great tool: https://github.com/Randgalt/record-builder
2. Nice article here about records: https://www.baeldung.com/java-record-keyword
3. Another, more in depth article: https://alidg.me/blog/2020/2/9/java14-records-in-depth
4. JEP link from above: https://openjdk.java.net/jeps/359

**Comment by pbludov** [comment]
Fix is merged. @nmancus1 should this issue be closed in favor of #8267?

**Comment by nmancus1** [comment]
@pbludov yes, please do.
